File: mobile/src/main/java/info/papdt/express/helper/ui/SettingsActivity.java
package info.papdt.express.helper.ui;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;

import info.papdt.express.helper.R;
import info.papdt.express.helper.support.Settings;
import info.papdt.express.helper.ui.common.AbsActivity;
import info.papdt.express.helper.ui.fragment.settings.*;

public class SettingsActivity extends AbsActivity {

	private int flag;

	private static final String EXTRA_SETTINGS_FLAG = "extra_flag";

	public static final int FLAG_MAIN = 0, FLAG_UI = 1, FLAG_LICENSE = 2, FLAG_NETWORK = 3,
			FLAG_CONTRIBUTORS = 4, FLAG_AUTO_DETECT = 5;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		if (getSettings().getBoolean(Settings.KEY_NAVIGATION_TINT, true)
				&& Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && !isNightMode()) {
			getWindow().setNavigationBarColor(getResources().getColor(R.color.colorPrimaryDark));
		}

		Intent intent = getIntent();
		flag = intent.getIntExtra(EXTRA_SETTINGS_FLAG, FLAG_MAIN);

		setContentView(R.layout.activity_settings);
	}

	@Override
	protected void setUpViews() {
		mActionBar.setDisplayHomeAsUpEnabled(true);

		Fragment f;
		switch (flag) {
			case FLAG_MAIN:
				f = new SettingsMain();
				mActionBar.setTitle(R.string.activity_settings);
				break;
			case FLAG_UI:
				f = new SettingsUi();
				mActionBar.setTitle(R.string.category_user_interface);
				break;
			case FLAG_LICENSE:
				getFragmentManager().beginTransaction().replace(R.id.container, new SettingsLicense())
						.commit();
				mActionBar.setTitle(R.string.open_source_license);
				return;
			case FLAG_NETWORK:
				f = new SettingsNetwork();
				mActionBar.setTitle(R.string.category_network);
				break;
			case FLAG_CONTRIBUTORS:
				f = new SettingsContributors();
				mActionBar.setTitle(R.string.category_contributors);
				break;
			case FLAG_AUTO_DETECT:
				f = new SettingsAutoDetect();
				mActionBar.setTitle(R.string.category_auto_detect_services);
				break;
			default:
				throw new RuntimeException("Please set flag when launching activity.");
		}
		getSupportFragmentManager().beginTransaction().replace(R.id.container, f).commit();
	}

	public static void launch(AppCompatActivity activity, int flag) {
		Intent intent = new Intent(activity, SettingsActivity.class);
		intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
		intent.putExtra(EXTRA_SETTINGS_FLAG, flag);
		activity.startActivity(intent);
	}

	public Snackbar makeSnackbar(String message, int duration) {
		return Snackbar.make($(R.id.container), message, duration);
	}

}

File: mobile/src/main/kotlin/info/papdt/express/helper/ui/SettingsActivity.kt
package info.papdt.express.helper.ui

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.View
import com.google.android.material.snackbar.Snackbar
import androidx.fragment.app.Fragment

import info.papdt.express.helper.R
import info.papdt.express.helper.support.ResourcesUtils
import info.papdt.express.helper.ui.common.AbsActivity
import info.papdt.express.helper.ui.fragment.settings.*
import info.papdt.express.helper.view.SwipeBackCoordinatorLayout
import moe.feng.kotlinyan.common.lazyFindNonNullView
import org.jetbrains.anko.withAlpha

class SettingsActivity : AbsActivity() {

	private lateinit var action: String

	private val coordinatorLayout: SwipeBackCoordinatorLayout
			by lazyFindNonNullView(R.id.swipe_back_coordinator_layout)
	private val rootLayout: View by lazyFindNonNullView(R.id.root_layout)
	private val rootViewBgColor by lazy {
		ResourcesUtils.getColorIntFromAttr(theme, R.attr.rootViewBackgroundColor)
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		action = intent.action ?: ACTION_MAIN

		setContentView(R.layout.activity_settings)
	}

	override fun setUpViews() {
		val f: Fragment
        mActionBar!!.setTitle(R.string.activity_settings)
		when (action) {
			ACTION_MAIN -> {
				f = SettingsMain()
			}
			ACTION_UI -> {
				f = SettingsUi()
			}
			ACTION_PUSH -> {
				f = SettingsPush()
			}
			ACTION_API -> {
				f = SettingsApi()
			}
			ACTION_ASSISTANT -> {
				f = SettingsAssistant()
			}
			ACTION_LICENSES -> {
				supportFragmentManager.beginTransaction().replace(R.id.container, SettingsLicense()).commit()
				mActionBar!!.setTitle(R.string.open_source_license)
				return
			}
			ACTION_CONTRIBUTORS -> {
				f = SettingsContributors()
			}
			else -> throw RuntimeException("Please set a supported action.")
		}
		supportFragmentManager.beginTransaction().replace(R.id.container, f).commit()

		if (f is SwipeBackCoordinatorLayout.OnSwipeListener) {
			coordinatorLayout.setOnSwipeListener(OnSwipeListener(f))
		}
	}

	fun makeSnackbar(message: String, duration: Int): Snackbar {
		return Snackbar.make(findViewById(R.id.container), message, duration)
	}

	inner class OnSwipeListener(listener: SwipeBackCoordinatorLayout.OnSwipeListener)
		: SwipeBackCoordinatorLayout.OnSwipeListener by listener {

		override fun onSwipeProcess(percent: Float) {
			rootLayout.setBackgroundColor(
					rootViewBgColor.withAlpha(
							(SwipeBackCoordinatorLayout.getBackgroundAlpha(percent) * 255)
									.toInt()
					)
			)
		}

        override fun onSwipeFinish(dir: Int) {
            window.statusBarColor = Color.TRANSPARENT
            finish()
        }

	}

	companion object {

		const val ACTION_MAIN = Intent.ACTION_MAIN
		const val ACTION_UI = "info.papdt.express.helper.action.SETTINGS_UI"
		const val ACTION_PUSH = "info.papdt.express.helper.action.SETTINGS_PUSH"
		const val ACTION_API = "info.papdt.express.helper.action.SETTINGS_API"
		const val ACTION_ASSISTANT = "info.papdt.express.helper.action.SETTINGS_ASSISTANT"
		const val ACTION_CONTRIBUTORS = "info.papdt.express.helper.action.SETTINGS_CONTRIBUTORS"
		const val ACTION_LICENSES = "info.papdt.express.helper.action.SETTINGS_LICENSES"

		fun launch(activity: Activity, action: String = ACTION_MAIN) {
			val intent = Intent(activity, SettingsActivity::class.java)
			intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
			intent.action = action
			activity.startActivity(intent)
		}

	}

}
